rosbag2_cpp/src/rosbag2_cpp/converter.cpp

```diff
@@ -17,66 +17,177 @@
 
+namespace
+{
+// Convenience struct to keep both type supports (rmw and introspection) together.
+struct ConverterTypeSupport
+{
+  std::shared_ptr<rcpputils::SharedLibrary> type_support_library;
+  const rosidl_message_type_support_t * rmw_type_support;
+
+  std::shared_ptr<rcpputils::SharedLibrary> introspection_type_support_library;
+  const rosidl_message_type_support_t * introspection_type_support;
+};
+
+template<typename T>
+T get_function_from(const char * function_name, std::shared_ptr<rcpputils::SharedLibrary> library)
+{
+  if (!library->has_symbol(function_name)) {
+    std::stringstream ss;
+    ss << "Converter could not find expected symbol '" << function_name <<
+      "' in rmw implementation " << library->get_library_path();
+    throw std::runtime_error{ss.str()};
+  }
+  T loaded_function = nullptr;
+  // Function expected to return a value because has_symbol was checked first.
+  loaded_function = (decltype(loaded_function))(library->get_symbol(function_name));
+  return loaded_function;
+}
+}  // namespace
+
+
 namespace rosbag2_cpp
 {
 
+typedef std::shared_ptr<rosbag2_storage::SerializedBagMessage> Msg;
+typedef std::shared_ptr<const rosbag2_storage::SerializedBagMessage> ConstMsg;
+
+class ConverterImpl
+{
+public:
+  explicit ConverterImpl(const ConverterOptions & options)
+  {
+    const std::string current_implementation_format{rmw_get_serialization_format()};
+    if (current_implementation_format == options.input_serialization_format) {
+      deserialize_fn_ = &rmw_deserialize;
+      ROSBAG2_CPP_LOG_INFO("The current impl is the DESER");
+    }
+    if (current_implementation_format == options.output_serialization_format) {
+      serialize_fn_ = &rmw_serialize;
+      ROSBAG2_CPP_LOG_INFO("The current impl is the SER");
+    }
+
+    auto packages_with_prefixes = ament_index_cpp::get_resources("rmw_typesupport");
+
+    for (const auto & package_prefix_pair : packages_with_prefixes) {
+      ROSBAG2_CPP_LOG_ERROR_STREAM("YOW A PAKAJ " << package_prefix_pair.first);
+      if (serialize_fn_ && deserialize_fn_) {
+        break;
+      }
+
+      const auto & pkg = package_prefix_pair.first;
+      if (pkg == "rmw_implementation") {
+        continue;
+      }
+
+      const auto libpath = rcpputils::find_library_path(pkg);
+      if (libpath.empty()) {
+        ROSBAG2_CPP_LOG_ERROR_STREAM("COULD NOT FIND LIB FOR " << pkg);
+        continue;
+      }
+      auto library = std::make_shared<rcpputils::SharedLibrary>(libpath);
+
+      auto get_format_fn = get_function_from<decltype(&rmw_get_serialization_format)>(
+        "rmw_get_serialization_format", library);
+      const char * fmt = get_format_fn();
+      if (!serialize_fn_ && fmt == options.input_serialization_format) {
+        deserialize_fn_ = get_function_from<decltype(deserialize_fn_)>(
+          "rmw_deserialize", library);
+      } else if (!deserialize_fn_ && fmt == options.output_serialization_format) {
```

Comment:
It looks like could be situation when condition `(!serialize_fn_ && fmt == options.input_serialization_format)` evaluates to `true` but we still need to process condition from else `if (!deserialize_fn_ && fmt ==options.output_serialization_format)`. \
Since `deserialize_fn_` also could be `null_ptr` at this moment. \
i.e. I think it would be more correct to move second `if` condition out from `else`